src/Scale.hs

```diff
@@ -70,126 +71,160 @@ data BaseMode
-baseModeIntervals DoubleHarmonicMajor =
-  zipToIntervalSet
-  [Perfect, Minor, Major, Perfect, Perfect, Minor, Major] [1..7]
-
-
-modeToIntervals :: Mode -> Set Interval
-modeToIntervals (Mode baseMode exts) = foldr extIntervals (baseModeIntervals baseMode) exts
-  where 
+zipToIntervalSet :: [Quality] -> [Int] -> Maybe (Set Interval)
+zipToIntervalSet quals sizes =
+  do ints <- zipWithM intervalFrom quals sizes
+     return $ fromList ints
+
+
+baseModeIntervals :: BaseMode -> Maybe (Set Interval)
+baseModeIntervals bm = if fromScratch then
+                         zipToIntervalSet bmQualities [1 .. 7]
+                       else
+                         do let (mode, shift) = modeAndShift
+                            ints <- baseModeIntervals mode
+                            nthDegreeIntervals ints shift
+  where
+    -- Discriminate between BaseModes for which we build the intervals from
+    -- scratch and those that are computed from some other interval set
+    fromScratch :: Bool
+    fromScratch = bm `elem` [ Ionian
+                            , AugmentedQuality
+                            , DiminishedQuality
+                            , MelodicMinor
+                            , HarmonicMinor
+                            , DoubleHarmonicMinor
+                            , HarmonicMajor
+                            , DoubleHarmonicMajor
+                            ]
+
+    -- The interval qualities for the modal interval sets built from scratch
+    bmQualities :: [Quality]
+    bmQualities =
+      case bm of
+        Ionian ->
+          [Perfect, Major, Major, Perfect, Perfect, Major, Major]
+        AugmentedQuality ->
+          [Perfect, Major, Major, Augmented 1, Augmented 1, Major, Minor]
+        DiminishedQuality ->
+          [Perfect, Major, Minor, Perfect, Diminished 1, Minor, Diminished 1]
+        MelodicMinor ->
+          [Perfect, Major, Minor, Perfect, Perfect, Major, Major]
+        HarmonicMinor ->
+          [Perfect, Major, Minor, Perfect, Perfect, Minor, Major]
+        DoubleHarmonicMinor ->
+          [Perfect, Major, Minor, Augmented 1, Perfect, Minor, Major]
+        HarmonicMajor ->
+          [Perfect, Major, Major, Perfect, Perfect, Minor, Major]
+        DoubleHarmonicMajor ->
+          [Perfect, Minor, Major, Perfect, Perfect, Minor, Major]
+
+    -- The starting mode and shift for modal interval sets built from other
+    -- interval sets
+    modeAndShift :: (BaseMode, Int)
+    modeAndShift =
+      case bm of
+        Dorian      -> (Ionian, 2)
+        Phrygian    -> (Ionian, 3)
+        Lydian      -> (Ionian, 4)
+        Mixolydian  -> (Ionian, 5)
+        Aeolian     -> (Ionian, 6)
+        Locrian     -> (Ionian, 7)
+        LydianAug   -> (MelodicMinor, 3)
+        LydianDom   -> (MelodicMinor, 4)
+        Altered     -> (MelodicMinor, 7)
+        PhrygianDom -> (HarmonicMinor, 5)
+
+
+modeToIntervals :: Mode -> Maybe (Set Interval)
```

Comment:
I've got the same question for this that I did for the Lib module -- can we assume or assure a guarantee that our Mode types always return a valid set of intervals? I think it'd make sense for that to be true

Reply:
Yeah, just like above I think the `Mode` type itself, if this function is implemented correctly, assures us that nothing will go wrong.